doc/threading_model.md

```diff
@@ -0,0 +1,31 @@
+﻿# gRPC Threading Model
+
+## Introduction
+This document serves as a basic intro for the gRPC threading model. gRPC generates service code in two fashions - synchronous and asynchronous API. These two closely correlate to the threading model in play. 
+
+## Synchronous API Threading model
+This is the easiest model to start with and could be a satisfactory model for a large number of services. As such, the implementation is efficient (and is not a ‘toy’ implementation). In this model, gRPC takes care of the threading completely and the application code need to worry only about handling of the actual rpc call.
+* When the server starts up, a thread pool is created to handle incoming rpc requests. The default implementation of this thread pool is called DynamicThreadPool. It creates threads equal to number of processors in the system and all of them are in waiting state.
+* DynamicThreadPool has in­built scaling mechanism to create additional threads on demand. For integrators interested in having a little more control on the behavior of the thread pool, they can define GRPC_CUSTOM_DEFAULT_THREAD_POOL and provide the implementation of the thread pool interface (which is literally just 1 function that adds work to the pool).

```

Comment:
`DynamicThreadPool` is no longer used for handling requests. Instead we use the `ThreadManager`. A `ThreadManager` internally maintains a thread pool and can be configured with a min and max number of polling threads.  This ensures that there are threads that are always waiting for incoming requests and also increases parallelism.

By default a sync server automatically creates as many server completion queues  (i.e completion queues used for notify about incoming RPC requests) as the number of cores on the machine it runs and one `ThreadManager` per completion queue.

#### Here are some pointers to the code:
-  [ThreadManager](https://github.com/grpc/grpc/tree/0a94f3c8ab55dfd12c14058d57f33121c8d6c411/src/cpp/thread_manager)  : Do read `thread_manager.h` interface. It has good comments. Note that `PollForWork` and `DoWork` are virtual functions. The implementation is in the [`SyncRequestThreadManager`](https://github.com/grpc/grpc/blob/0a94f3c8ab55dfd12c14058d57f33121c8d6c411/src/cpp/server/server_cc.cc#L260) class in `server_cc.cc` .
- Some relevant code in `ServerBuilder` (where defaults are set):   [1](https://github.com/grpc/grpc/blob/34011f19ed40503fa658234742b3a4b85476af1a/include/grpc%2B%2B/server_builder.h#L212),  [2](https://github.com/grpc/grpc/blob/0a94f3c8ab55dfd12c14058d57f33121c8d6c411/src/cpp/server/server_builder.cc#L59),  [3](https://github.com/grpc/grpc/blob/0a94f3c8ab55dfd12c14058d57f33121c8d6c411/src/cpp/server/server_builder.cc#L273) and  [4](https://github.com/grpc/grpc/blob/0a94f3c8ab55dfd12c14058d57f33121c8d6c411/src/cpp/server/server_builder.cc#L273)
- Relevant code in `Server` : [1](https://github.com/grpc/grpc/blob/0a94f3c8ab55dfd12c14058d57f33121c8d6c411/src/cpp/server/server_cc.cc#L368), [2](https://github.com/grpc/grpc/blob/0a94f3c8ab55dfd12c14058d57f33121c8d6c411/src/cpp/server/server_cc.cc#L389) and [3](https://github.com/grpc/grpc/blob/0a94f3c8ab55dfd12c14058d57f33121c8d6c411/src/cpp/server/server_cc.cc#L260)

Hope this helps.